src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ContextAware.cs

```diff
@@ -0,0 +1,95 @@
+{
+    internal sealed class ContextAwareDataContractIndex
+    {
+        private (DataContract? strong, WeakReference<DataContract>? weak)[] _contracts;
+        private ConditionalWeakTable<Type, DataContract> _keepAlive;
+
+        public int Length => _contracts.Length;
+
+        public ContextAwareDataContractIndex(int size)
+        {
+            _contracts = new (DataContract?, WeakReference<DataContract>?)[size];
+            _keepAlive = new ConditionalWeakTable<Type, DataContract>();
+        }
+
+        public DataContract? GetItem(int index) => _contracts[index].strong ?? (_contracts[index].weak?.TryGetTarget(out DataContract? ret) == true ? ret : null);
+
+        public void SetItem(int index, DataContract dataContract)
+        {
+            // Check for unloadability to decide how to store the value
+            AssemblyLoadContext? alc = AssemblyLoadContext.GetLoadContext(dataContract.UnderlyingType.Assembly);
+            if (alc == null || !alc.IsCollectible)
+            {
+                _contracts[index].strong = dataContract;
+            }
+            else
+            {
+                _contracts[index].weak = new WeakReference<DataContract>(dataContract);
+                _keepAlive.Add(dataContract.UnderlyingType, dataContract);
+            }
+        }
+
+        public void Resize(int newSize)
+        {
+            Array.Resize<(DataContract?, WeakReference<DataContract>?)>(ref _contracts, newSize);
+        }
+    }
+
+    internal sealed class ContextAwareDictionary<TKey, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] TValue>
+        where TKey : Type
+        where TValue : class?
+    {
+        private readonly ConcurrentDictionary<TKey, TValue> _fastDictionary = new();
+        private readonly ConditionalWeakTable<TKey, TValue> _collectibleTable = new();
+
+
+        internal TValue GetOrAdd(TKey t, Func<TKey, TValue> f)
+        {
+            TValue? ret;
+
+            // The fast and most common default case
+            if (_fastDictionary.TryGetValue(t, out ret))
+                return ret;
+
+            // Common case for collectible contexts
+            if (_collectibleTable.TryGetValue(t, out ret))
+                return ret;
+
+            // Not found. Do the slower work of creating the value in the correct collection.
+            AssemblyLoadContext? alc = AssemblyLoadContext.GetLoadContext(t.Assembly);
+
+            // Null and non-collectible load contexts use the default table
+            if (alc == null || !alc.IsCollectible)
+            {
+                return _fastDictionary.GetOrAdd(t, f);
+            }
+
+            // Collectible load contexts should use the ConditionalWeakTable so they can be unloaded
+            else
+            {
+                lock (_collectibleTable)
+                {
+                    if (!_collectibleTable.TryGetValue(t, out ret))
+                    {
+                        ret = f(t);
```

Comment:
The delegate isn't necessarily a simple amount of work depending on your OM design. It's something you really want to avoid executing multiple times. For example, if you instantiate a DCS on an incoming request on a REST api for example, having 100 concurrent requests could result in redoing this expensive work 100 times. You could negatively impact your first request time significantly.  
Can you provide me information about it being prone to deadlocks? The code run by the delegate isn't going to do anything async so any reentrance will occur on the same thread. I don't believe this specific scenario is able to deadlock, but I'm open to learning about ways it can deadlock that I might not be aware of.

Reply:
You know better if you think that creating a DCS will not execute arbitrary code. But note that [`ConcurrentDictionary.GetOrAdd` will not execute the delegate inside a lock](https://github.com/dotnet/runtime/blob/7980421b2c2a8cc21c2060f831b8020ef0517632/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1181). In the most common case of non-unloadable assemblies there is still the possibility that the delegate will run many times.

There is also [`ConditionalWeakTable.CreateValue`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2.getvalue) that you can use like `Concurrent.Dictionary.GetOrAdd`.

Reply:
I agree, we should place a lock around ConcurrentDictionary.GetOrAdd. We don't need to check a second time if we are always holding a lock when adding as that would guarantee the prior add has completed before calling GetOrAdd and it will act like a Get.